### PR TITLE
fix: refreshing tree upon config change or kubeconfig removal (#595)

### LINKED ISSUE #595: Logging into the cluster does not refresh the tree
depends on https://github.com/redhat-developer/intellij-common/issues/201

**Steps:**
1. ASSERT: make sure that you are not logged into your current cluster
2. EXEC: launch the plugin, unfold the root (cluster) node
3. ASSERT: tree shows "Please log into the cluster"
4. EXEC: pick "Log into the cluster" in the context menu and log in
5. ASSERT: dialog is closed, look at the tree

**Result:**
The tree still shows "Please log in to the cluster". "Refresh" doesnt help, I have to restart the IDE. 
image

### LINKED ISSUE #621: If my cluster is not reachable I get a "Fatal IDE error"
**Steps:**
1. ASSERT: make sure that the cluster for the current context is not reachable
2. EXEC: launch IDE and open OpenShift toolkit window (you can also, if you have 2 contexts, switch to the unreachable context using intellij-kubernetes)

**Result:**
you get a "Fatal IDE error" reported.
The stacktrace is something similar to the following:
```java
java.io.IOException: Process exited with an error: 1 (Exit value: 1) {
	"message": "unable to list all dynamic resources required to find components: failed to fetch api groups from kubernetes: Get \"https://api.akundu-221120231226.devcluster.openshift.com:6443/api?timeout=32s\": dial tcp: lookup api.akundu-221120231226.devcluster.openshift.com on 10.2.0.1:53: no such host"
}

	at com.redhat.devtools.intellij.common.utils.ExecHelper.executeWithResult(ExecHelper.java:261)
	at org.jboss.tools.intellij.openshift.utils.odo.OdoCli.execute(OdoCli.java:280)
	at org.jboss.tools.intellij.openshift.utils.odo.OdoCli.discover(OdoCli.java:891)
	at org.jboss.tools.intellij.openshift.tree.application.ApplicationRootNodeOdo.discover(ApplicationRootNodeOdo.java:275)
	at org.jboss.tools.intellij.openshift.tree.application.ApplicationsRootNode.addContext(ApplicationsRootNode.java:180)
	at org.jboss.tools.intellij.openshift.tree.application.ApplicationsRootNode.loadProjectModel(ApplicationsRootNode.java:134)
	at org.jboss.tools.intellij.openshift.tree.application.ApplicationsRootNode.lambda$getOdo$1(ApplicationsRootNode.java:85)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2325)
	at org.jboss.tools.intellij.openshift.tree.application.ApplicationsRootNode.getOdo(ApplicationsRootNode.java:85)
	at org.jboss.tools.intellij.openshift.tree.application.ApplicationsRootNode.getOdo(ApplicationsRootNode.java:92)
	at org.jboss.tools.intellij.openshift.tree.application.ApplicationsTreeStructure.getChildElements(ApplicationsTreeStructure.java:71)
	at com.intellij.ui.tree.StructureTreeModel.getValidChildren(StructureTreeModel.java:411)
	at com.intellij.ui.tree.StructureTreeModel.validateChildren(StructureTreeModel.java:329)
	at com.intellij.ui.tree.StructureTreeModel.getNode(StructureTreeModel.java:323)
	at com.intellij.ui.tree.StructureTreeModel.getChildren(StructureTreeModel.java:343)
	at com.intellij.ui.tree.AsyncTreeModel$CmdGetChildren.getNode(AsyncTreeModel.java:568)
	at com.intellij.ui.tree.AsyncTreeModel$Command.get(AsyncTreeModel.java:464)
	at com.intellij.ui.tree.AsyncTreeModel$Command.get(AsyncTreeModel.java:430)
	at com.intellij.util.concurrency.Invoker$Task.run(Invoker.java:306)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1133)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runInReadActionWithWriteActionPriority$0(ProgressIndicatorUtils.java:76)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtilService.runActionAndCancelBeforeWrite(ProgressIndicatorUtilService.java:63)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runActionAndCancelBeforeWrite(ProgressIndicatorUtils.java:133)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runWithWriteActionPriority$1(ProgressIndicatorUtils.java:114)
	at com.intellij.openapi.progress.ProgressManager.lambda$runProcess$0(ProgressManager.java:71)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:186)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:604)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:679)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:635)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:603)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:61)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:173)
	at com.intellij.openapi.progress.ProgressManager.runProcess(ProgressManager.java:71)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runWithWriteActionPriority(ProgressIndicatorUtils.java:111)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runInReadActionWithWriteActionPriority(ProgressIndicatorUtils.java:76)
	at com.intellij.util.concurrency.Invoker.invokeSafely(Invoker.java:188)
	at com.intellij.util.concurrency.Invoker.lambda$offerSafely$0(Invoker.java:165)
	at com.intellij.util.concurrency.Invoker$Background.lambda$offer$0(Invoker.java:433)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:249)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$200(BoundedTaskExecutor.java:31)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.executeFirstTaskAndHelpQueue(BoundedTaskExecutor.java:227)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:218)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.run(BoundedTaskExecutor.java:215)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)
	at org.apache.commons.exec.DefaultExecutor.executeInternal(DefaultExecutor.java:404)
	at org.apache.commons.exec.DefaultExecutor.execute(DefaultExecutor.java:166)
	at com.redhat.devtools.intellij.common.utils.ExecHelper.executeWithResult(ExecHelper.java:258)
	... 51 more
```

### PR DESCRIPTION
**Pull request by adietish**
depends on https://github.com/redhat-developer/intellij-common/issues/202
fixes #595
fixes #621

### CODE REVIEW

**Comment by adietish** [comment]
@sbouchet: This is ready for review as soon as https://github.com/redhat-developer/intellij-common/pull/202 is merged and published.
It makes sure that the tree is refreshed as soon as there's a change in config or that the kube config was removed (and also when it reappears)

**Comment by sbouchet** [comment]
@adietish : for the two uses cases it works well.
1- after loging to devsandbox
2 - after removal of kube config file.

But i got an issue when i switch clusters on kubernetes view. see the screencast
[Capture vidéo du 2023-11-16 16-32-16.webm](https://github.com/redhat-developer/intellij-openshift-connector/assets/600414/0e805517-0928-4c3e-9a77-5a7a7c4a7da1)

look at when i switch cluster, i got error on helm charts.
trying another time, i can see clusters again, but on openshift i got default cluster name ( kubernetes.default.svc )

**Comment by adietish** [comment]
@sbouchet: can you please verify the following:
When you face the case where the current namespace that is displayed in openshift tooling is wrong, do you have a current namespace set?
In the cases where I can replicate your problem statement, I had the current namespace not set. Wondering if you can confirm this. Thanks.

**Comment by adietish** [comment]
@sbouchet: The error where the helm charts could not be fetched any more went away for me once I depended on intellij-kubernetes 1.2.0.271. Can you please verify?

**Comment by adietish** [comment]
added a commit that fixes #621

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliTest.java [hunk 1]
```diff
@@ -81,8 +81,8 @@ protected void tearDown() throws Exception {
     protected void createProject(String project) throws IOException, ExecutionException, InterruptedException {
         odo.createProject(project);
         // need to refresh kubernetes client with the correct namespace
-        ToolFactory.getInstance().resetOdo();
-        odo = ToolFactory.getInstance().getOdo(getProject()).get();
+        //resetOdo();
```
Comment (sbouchet) [inline src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliTest.java:84]: remove this line completely

File: src/main/java/org/jboss/tools/intellij/openshift/tree/application/BaseNode.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Red Hat, Inc.
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/tree/application/BaseNode.java:2]: should be 2023

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 3]
```diff
@@ -862,7 +867,12 @@ public String consoleURL() throws IOException {
 
     @Override
     public boolean isOpenShift() {
-        return ClusterHelper.getClusterInfo(client).isOpenshift();
+        OpenShiftClient osClient = client.adapt(OpenShiftClient.class);
+        try {
+            return osClient.isSupported();
+        } catch (KubernetesClientException e) {
+            return e.getCode() == HttpURLConnection.HTTP_UNAUTHORIZED;
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:874]: so if adapt to OpenshiftClient fails with HTTP_UNAUTHORIZED , it means it's an openshift cluster ?

**Review by sbouchet — COMMENTED** [review]
please review comments

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java (hunk 3 above)
Comment (adietish) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:874]: Yes, I have the same in intellij-kubernetes.

**Comment by adietish** [comment]
> please review comments

fixed your complaints. Please re-review.

**Review by sbouchet — APPROVED** [review]
reviewed with local minikube and devsandbox. LGTM

**Pull request merged**